lalrpop/src/normalize/precedence/mod.rs

```diff
@@ -0,0 +1,334 @@
+    // After prevalidation, either each or none of the alternative of a nonterminal have precedence
+    // annotations, so we just have to check the first one.
+    non_term.alternatives.first().map(|alt|
+        alt.annotations.iter().any(|ann|
+            ann.id == Atom::from(PREC_ANNOT)
+            || ann.id == Atom::from(ASSOC_ANNOT)
+        )
+    )
+    .unwrap_or(false)
+}
+
+/// Expand a rule with precedence annotations. As it implies to generate new rules, return a vector
+/// of grammar items.
+fn expand_nonterm(mut nonterm: NonterminalData) -> NormResult<Vec<GrammarItem>> {
+    let alt_with_ann = Vec::with_capacity(nonterm.alternatives.len());
+
+    let (mut lvls, alts_with_ann) = mem::replace(&mut nonterm.alternatives, vec![]).into_iter().fold(
+        (vec![], alt_with_ann),
+        |(mut lvls, mut acc): (Vec<u32>, Vec<(u32, Option<Assoc>, Alternative)>), mut alt| {
+            // All the following unsafe `unwrap()`, `panic!()`, etc. should never panic thanks to
+            // prevalidation. Prevalidation must ensure that each alternative is annotated with at
+            // least a precedence level, each precedence annotation must have an argument which
+            // is parsable as an integer, and each optional assoc annotation must have an argument
+            // that is either "right", "left" or "none".
+
+            // Extract and remove precedence and associativity annotations
+            let lvl: u32 = {
+                let index = alt.annotations.iter().position(|ann| ann.id == Atom::from(PREC_ANNOT)).unwrap();
+                let (_, val) = alt.annotations.remove(index).arg.unwrap();
+                val.parse().unwrap()
+            };
+            let assoc: Option<Assoc> = alt.annotations.iter().position(|ann| ann.id == Atom::from(ASSOC_ANNOT)).map(|index| {
+                let (_, val) = alt.annotations.remove(index).arg.unwrap();
+                val.parse().unwrap()
+            });
+
+            acc.push((lvl, assoc, alt));
+            lvls.push(lvl);
+            (lvls, acc)
+        });
+
+    lvls.sort();
+    lvls.dedup();
+
+    let rest = &mut alts_with_ann.into_iter();
+
+    let lvl_max = *lvls.last().unwrap();
+    let result = lvls.into_iter().map(|lvl| {
+        // The generated non terminal corresponding to the last level keeps the same name as the
+        // initial item, so that all external references to it are still valid. Other levels get
+        // the names `Name1`, `Name2`, etc. where `Name` is the name of the initial item.
+        let name = NonterminalString(Atom::from(
+            if lvl == lvl_max {
+                format!("{}", nonterm.name)
+            }
+            else {
+                format!("{}{}", nonterm.name, lvl)
+            }
+        ));
+
+        let nonterm_prev = if lvl > 1 {
+                Some(SymbolKind::Nonterminal(NonterminalString(Atom::from(
+                format!("{}{}", nonterm.name, lvl - 1)))))
```

Comment:
I might be wrong, but the levels does not need to be consecutive, right? So this would need to look to the previous level in the `lvls` Vec

Reply:
Looks like this is validated earlier at least, but it seems pretty arbitrary to limit the levels to be consecutive?

Reply:
Yeah, they don't need to. My initial idea was that maybe if there's a hole, this can be a copy/paste error from the user. On the other hand, it's probably less ergonomic, as if one wants to delete one level, he has to renumber everything from there. I can remove this restriction if you wish.

Reply:
Removed the requirement for consecutive levels from `1` to `n` in a374f6d and 4c41182.